src/charm.py

```diff
@@ -113,85 +106,125 @@ def _on_jenkins_pebble_ready(self, event: PebbleReadyEvent) -> None:
             return
 
         self.unit.set_workload_version(version)
-        self.unit.status = ActiveStatus()
+        self.unit.status = ops.ActiveStatus()
 
-    def _on_get_admin_password(self, event: ActionEvent) -> None:
+    def _on_get_admin_password(self, event: ops.ActionEvent) -> None:
         """Handle get-admin-password event.
 
         Args:
             event: The event fired from get-admin-password action.
         """
-        if not self._jenkins_container.can_connect():
+        container = self.unit.get_container(self.state.jenkins_service_name)
+        if not container.can_connect():
             event.defer()
             return
-        credentials = jenkins.get_admin_credentials(self._jenkins_container)
+        credentials = jenkins.get_admin_credentials(container)
         event.set_results({"password": credentials.password})
 
-    def _on_update_status(self, _: UpdateStatusEvent) -> None:
-        """Handle update status event.
+    def _remove_unlisted_plugins(self, container: ops.Container) -> ops.StatusBase:
+        """Remove plugins that are installed but not allowed.
 
-        On Update status:
-        1. Update Jenkins patch LTS version if available.
-        2. Update apt packages if available.
+        Args:
+            container: The jenkins workload container.
+
+        Returns:
+            The unit status of the charm after the operation.
         """
-        if self.state.update_time_range and not self.state.update_time_range.check_now():
-            self.unit.status = ActiveStatus()
-            return
+        original_status = self.unit.status.name
+        try:
+            jenkins.remove_unlisted_plugins(plugins=self.state.plugins, container=container)
+        except (jenkins.JenkinsPluginError, jenkins.JenkinsError) as exc:
+            logger.error("Failed to remove unlisted plugin, %s", exc)
+            return ops.StatusBase.from_name(original_status, "Failed to remove unlisted plugin.")
+        except TimeoutError as exc:
+            logger.error("Failed to restart jenkins after removing plugin, %s", exc)
+            return ops.BlockedStatus("Failed to restart Jenkins (remove plugin)")
+        return ops.ActiveStatus()
+
+    def _update_jenkins_version(self, container: ops.Container) -> ops.StatusBase:
+        """Update Jenkins patch version if available.
 
-        self.unit.status = ActiveStatus("Checking for updates.")
+        The update will only take place if the current time is within the restart-time-range config
+        value.
+
+        Args:
+            container: The Jenkins workload container.
+
+        Returns:
+            The unit status of the charm after the operation.
+        """
+        original_status = self.unit.status.name
+        self.unit.status = ops.StatusBase.from_name(original_status, "Checking for updates.")
         try:
             latest_patch_version = jenkins.get_updatable_version(proxy=self.state.proxy_config)
```

Comment:
Did we discuss not returning the version and instead indicating whether Jenkins needs to be updated?

Reply:
Yup, I think it is better suited on a different refactor branch! I've got it separately marked as #TODO comment locally :)